### PR TITLE
MultiGoalShortestPath improvements

### PR DESCRIPTION
**Pull request by aclegg3**
## Motivation and Context

Currently, if any of the requested_end points don't snap to the navmesh, `MultiGoalShortestPath` will early abort and return `inf` distance.
Expected behavior: if *any* of the points in `requested_ends` are valid, the shortest path should be computed accurately.

Additional feature: adds `closest_end_point_index` to `MultiGoalShortestPath` struct API. Makes it easier to know which point form the input set was actually selected as closets. Corresponds to the end of the `path`.

## How Has This Been Tested

TODO:
May need to update tests to check these edge cases:
- all points off navmesh
- some points off navmesh
- all points on other island
- all point son other island or off navmesh

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x ] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/esp/bindings/ShortestPathBindings.cpp [hunk 1]
```diff
@@ -66,7 +66,11 @@ void initShortestPathBindings(py::module& m) {
           R"(A list of points that specify the shortest path on the navigation mesh between requestedStart and the closest (by geodesic distance) point in requestedEnds. Will be empty if no path exists.)")
       .def_readwrite(
           "geodesic_distance", &MultiGoalShortestPath::geodesicDistance,
-          R"(The total geodesic distance of the path. Will be inf if no path exists.)");
+          R"(The total geodesic distance of the path. Will be inf if no path exists.)")
+      .def_readwrite(
+          "closest_end_point_index",
+          &MultiGoalShortestPath::closestEndPointIndex,
+          R"(he index of the closest end point corresponding to end of the shortest path. Will be -1 if no path exists.)");
```
Comment (jturner65) [inline src/esp/bindings/ShortestPathBindings.cpp:73]: Slight spelling issue
`R"(he index of the closest end point corresponding to end of the shortest path. Will be -1 if no path exists.)");`

should be 
`R"(The index of the closest end point corresponding to end of the shortest path. Will be -1 if no path exists.)");`

File: src/esp/nav/PathFinder.cpp [hunk 2]
```diff
@@ -1349,18 +1350,26 @@ bool PathFinder::Impl::findPathSetup(MultiGoalShortestPath& path,
...
         projectToPoly(rqEnd, navQuery_.get(), filter_.get());
 
     if (status != DT_SUCCESS || endRef == 0) {
-      return false;
+      ESP_DEBUG() << "Can't project end-point to navmesh, skipping: " << rqEnd;
+    } else {
+      numValidPoints++;
     }
 
     path.pimpl_->endRefs.emplace_back(endRef);
```
Comment (jturner65) [inline src/esp/nav/PathFinder.cpp:1366]: Do you really want to potentially add an endRef == 0 to the store of endRefs?

Comment (jturner65) [inline src/esp/nav/PathFinder.cpp:1366]: Do you really want to potentially add an endRef == 0 to the store of endRefs?

File: src/esp/nav/PathFinder.cpp [hunk 3]
```diff
@@ -1349,18 +1350,26 @@ bool PathFinder::Impl::findPathSetup(MultiGoalShortestPath& path,
...
     path.pimpl_->pathEnds.emplace_back(pathEnd);
+
+    if (numValidPoints == 0) {
```
Comment (Skylion007) [inline src/esp/nav/PathFinder.cpp:1369]: Hold on, won't this still abort if the FIRST validPoint is doesn't snap even if subsequent ones are valid? Are the paths sorted?

File: src/esp/nav/PathFinder.cpp [hunk 4]
```diff
@@ -1349,18 +1350,26 @@ bool PathFinder::Impl::findPathSetup(MultiGoalShortestPath& path,
...
+  int numValidPoints = 0;
   for (const auto& rqEnd : path.getRequestedEnds()) {
     dtPolyRef endRef = 0;
     vec3f pathEnd;
     std::tie(status, endRef, pathEnd) =
         projectToPoly(rqEnd, navQuery_.get(), filter_.get());
 
     if (status != DT_SUCCESS || endRef == 0) {
-      return false;
+      ESP_DEBUG() << "Can't project end-point to navmesh, skipping: " << rqEnd;
```
Comment (Skylion007) [inline src/esp/nav/PathFinder.cpp:1361]: If it's really skipping, shouldn't it have a continue logic in the loop?

File: src/esp/nav/PathFinder.cpp (hunk 3 above)
Comment (Skylion007) [reply src/esp/nav/PathFinder.cpp:1369]: If it is sorted, add a comment.

Comment (aclegg3) [reply src/esp/nav/PathFinder.cpp:1369]: Lol, thank you for catching this. Typo, the conditional should be outside the loop.

File: src/esp/nav/PathFinder.cpp (hunk 4 above)
Comment (aclegg3) [reply src/esp/nav/PathFinder.cpp:1361]: Because the original end-points will be sorted in the next step, we still need an entry for invalid points so the indices will match. This warning message is primarily intended to let the user know that they are providing invalid points in case that is an unexpected bug in their algorithm. The skip here is implicit.

Comment (aclegg3) [reply src/esp/nav/PathFinder.cpp:1365]: Added a way to track this and skip more explicitly.

**Review by jturner65 — APPROVED** [review]
LGTM!

**Review by jturner65 — APPROVED** [review]

**Review by jturner65 — APPROVED** [review]
